model-optimizer/extensions/front/tf/MapFNTransformation.py

```diff
@@ -229,27 +202,46 @@ def transform_keras_rnn_output_concatenation(external_match: dict, internal_matc
-            # check connections between body output ports and external output ports of Loop node
-            if Loop.back_edge_exists(loop_node.back_edges, internal_match['concatenation_result'].internal_layer_id,
-                                     internal_match['container'].internal_layer_id) and \
-                    Loop.back_edge_exists(loop_node.back_edges,
-                                          internal_match['increment_iteration_result'].internal_layer_id,
-                                          internal_match['current_iteration'].internal_layer_id) and \
-                    Loop.inter_edge_exists(loop_node.input_port_map, reserve_port_idx,
-                                           internal_match['container'].internal_layer_id) and \
-                    Loop.inter_edge_exists(loop_node.output_port_map, stack_port_idx,
-                                           internal_match['concatenation_result'].internal_layer_id):
-                KerasRNNOutputConcatenation.transform_keras_rnn_output_concatenation(external_match, internal_match)
+    def find_and_replace_pattern(self, graph: Graph):
+        for loop_node in graph.get_op_nodes(op='Loop'):
+            body_graph = loop_node['body']
+            body_pattern = MapFNOutputConcatenation.get_body_pattern()
+            internal_matches = find_subgraph_match_to_pattern(body_graph, body_pattern)
+
+            for internal_match in internal_matches:
+                # check if TensorListReserve from the main graph is connected with Parameter node from the body graph
+                # that is assigned for storing intermediate output results of While Loop. If yes, the transformation
+                # detects intermediate outputs concatentation by this port and can use Loop axis attribute
+                reserve_node = Loop.get_external_nodes_by_internal_id(loop_node,
+                                                                     internal_match['container'].internal_layer_id)
+                reserve_node = reserve_node[0] if (len(reserve_node) == 1 and
+                                                   reserve_node[0].op == 'TensorListReserve') else None
+                if reserve_node is None:
+                    continue
+                stack_node = Loop.get_external_nodes_by_internal_id(loop_node,
+                                                                   internal_match[
+                                                                       'concatenation_result'].internal_layer_id)
+                stack_node = stack_node if len(stack_node) == 1 else None
+
+                # skip StopGradient node if it exists between While loop output port and TensorListStack operation
+                if stack_node is None:
+                    continue
+                if stack_node[0].op == 'StopGradient':
+                    stack_node = [dest.node for dest in stack_node[0].out_port(0).get_destinations()]
```

Comment:
I would recommend to skip all nodes which are marked with `identity=True` attribute in the while loop.  Use method `skip_nodes_by_condition` from the ObjectDetectionAPI.py file. You can move it to some common graph functions module.

Reply:
Done. Thanks.